File: python/tests/test_enumerators.py
"""Behavioural tests for the four enumerators, against the committed fixture.

Added in v0.9.6 to close a real gap: before this file, ``datalib_countries``,
``datalib_surveys``, ``datalib_vintages`` and ``datalib_adaptations`` -- four of
the thirteen canonical commands -- had **no behavioural coverage at all** in
either CI-runnable language. The only references anywhere in the suite were
name-surface lists in :mod:`test_docs` (that the function exists and is
exported), which cannot catch a wrong return value.

These are deliberately NOT shared golden cases. The three ports return
structurally different objects for the same query -- Stata's
``datalib_adaptations`` returns ``r(collections)="HLT IPUMS"`` and its
``datalib_vintages`` returns bare integers in ``r(masters)="9 10"``, R returns
data frames whose ``folder`` column holds full vintage folder names, and Python
returns ``VintageCatalog`` / ``AdaptationInfo`` objects. A single shared
``expect_items`` column could therefore match at most one language. Whether the
return shapes should converge is a contract question for
``config/grammar.md``, not something a CSV can paper over; until it is settled,
each language asserts its own shape against the same fixture tree.
"""

from __future__ import annotations

import pytest

import datalib as dl


def test_countries_are_uppercase_and_complete(fixture_root) -> None:
    assert dl.datalib_countries(root=fixture_root) == ["BRA", "KEN", "ZWE"]


def test_surveys_lists_the_survey_folder(fixture_root) -> None:
    assert dl.datalib_surveys("ZWE", root=fixture_root) == ["ZWE_2019_MICS"]


def test_country_is_uppercased_at_the_boundary(fixture_root) -> None:
    """Contract rule 1: inputs are trimmed and uppercased before matching."""
    assert dl.datalib_surveys("zwe", root=fixture_root) == ["ZWE_2019_MICS"]
    assert dl.datalib_surveys("  ZWE  ", root=fixture_root) == ["ZWE_2019_MICS"]


def test_vintages_reports_masters_as_integers(fixture_root) -> None:
    cat = dl.datalib_vintages("ZWE", 2019, "MICS", root=fixture_root)
    assert cat.masters == (1, 2)


def test_latest_master_is_the_numeric_maximum_not_alphabetical(fixture_root) -> None:
    """Contract rule 2: latest is the numeric maximum, so v10 > v09.

    BRA_2015_PNAD carries v09 and v10 precisely so this cannot pass by accident:
    directory-listing or alphabetical order would pick v09.
    """
    cat = dl.datalib_vintages("BRA", 2015, "PNAD", root=fixture_root)
    assert cat.masters == (9, 10)
    assert max(cat.masters) == 10


def test_adaptations_lists_collection_names(fixture_root) -> None:
    names = [a.name for a in dl.datalib_adaptations("ZWE", 2019, "MICS", root=fixture_root)]
    assert names == ["HLT", "IPUMS"]


def test_a_country_with_no_adaptations_returns_empty(fixture_root) -> None:
    assert dl.datalib_adaptations("BRA", 2015, "PNAD", root=fixture_root) == []


def test_present_but_empty_root_is_an_empty_listing(tmp_path) -> None:
    """A root that exists but holds no countries is empty, not an error.

    This is the half of enumerator emptiness the contract already settles; the
    absent-parent half is the divergence below.
    """
    assert dl.datalib_countries(root=tmp_path) == []


@pytest.mark.xfail(
    strict=True,
    reason=(
        "v0.10.0: Python must raise on an absent parent. Stata exits 198 and R "
        "raises datalib_error_input for an unknown country; Python returns []. "
        "Tracked in the config/grammar.md Alignment status table "
        "('enumerators error on an absent parent'). strict=True means this "
        "becomes a FAILURE the moment Python is fixed -- delete the marker then."
    ),
)
def test_unknown_country_raises(fixture_root) -> None:
    with pytest.raises(dl.errors.NotFoundError):
        dl.datalib_surveys("XXX", root=fixture_root)

File: r/tests/testthat/test-enumerators.R
# Behavioural tests for the four enumerators, against the committed fixture.
#
# Added in v0.9.6 to close a real gap: before this file, datalib_countries,
# datalib_surveys, datalib_vintages and datalib_adaptations -- four of the
# thirteen canonical commands -- had NO behavioural coverage at all in either
# CI-runnable language. The only references in the suite were the name-surface
# lists in test-docs.R (that the function exists and is documented), which
# cannot catch a wrong return value.
#
# These are deliberately NOT shared golden cases. The three ports return
# structurally different objects for the same query: Stata's datalib_vintages
# returns bare integers in r(masters) = "9 10" and its datalib_adaptations
# returns r(collections) = "HLT IPUMS", Python returns VintageCatalog /
# AdaptationInfo objects, and R (below) returns data frames whose `folder`
# column holds full vintage folder names. One shared expect_items column could
# match at most one of the three. Whether the shapes should converge is a
# contract question for config/grammar.md, not something a CSV can paper over.

fixture <- datalib_test_fixture_root()

test_that("datalib_countries returns the fixture countries, uppercased", {
  got <- datalib_countries(root = fixture)
  expect_s3_class(got, "data.frame")
  expect_identical(names(got), "country")
  expect_identical(got$country, c("BRA", "KEN", "ZWE"))
})

test_that("datalib_surveys lists the survey folder with its parts", {
  got <- datalib_surveys("ZWE", root = fixture)
  expect_identical(got$folder, "ZWE_2019_MICS")
  expect_identical(got$country, "ZWE")
  expect_identical(got$survey, "MICS")
})

test_that("country is uppercased and trimmed at the boundary (contract rule 1)", {
  expect_identical(datalib_surveys("zwe", root = fixture)$folder, "ZWE_2019_MICS")
  expect_identical(datalib_surveys("  ZWE  ", root = fixture)$folder, "ZWE_2019_MICS")
})

test_that("datalib_vintages reports masters and adaptations with parsed parts", {
  got <- datalib_vintages("ZWE", 2019, "MICS", root = fixture)
  expect_identical(got$folder,
                   c("ZWE_2019_MICS_v01_M",
                     "ZWE_2019_MICS_v02_M",
                     "ZWE_2019_MICS_v02_M_v01_A_HLT",
                     "ZWE_2019_MICS_v02_M_v01_A_IPUMS"))
  expect_identical(got$kind,
                   c("master", "master", "adaptation", "adaptation"))
  expect_identical(got$master_version, c(1L, 2L, 2L, 2L))
})

test_that("latest master is the numeric maximum, not alphabetical (contract rule 2)", {
  # BRA_2015_PNAD carries v09 and v10 precisely so this cannot pass by accident:
  # directory-listing or alphabetical order would pick v09.
  got <- datalib_vintages("BRA", 2015, "PNAD", root = fixture)
  expect_identical(got$master_version, c(9L, 10L))
  expect_identical(got$folder[got$is_latest], "BRA_2015_PNAD_v10_M")
  expect_identical(max(got$master_version), 10L)
})

test_that("datalib_adaptations lists collections under the chosen master", {
  got <- datalib_adaptations("ZWE", 2019, "MICS", root = fixture)
  expect_identical(got$collection, c("HLT", "IPUMS"))
  expect_identical(got$master_version, c(2L, 2L))
})

test_that("a country with no adaptations returns zero rows, not an error", {
  got <- datalib_adaptations("BRA", 2015, "PNAD", root = fixture)
  expect_identical(nrow(got), 0L)
})

test_that("an unknown country is a typed input error", {
  # R already conforms here; Stata exits 198 and Python still returns empty
  # (tracked in the config/grammar.md Alignment status table, and marked
  # xfail(strict) in python/tests/test_enumerators.py).
  expect_error(datalib_surveys("XXX", root = fixture),
               class = "datalib_error_input")
})

test_that("a root that exists but holds no countries is an empty listing", {
  tmp <- file.path(tempdir(), "datalib-empty-root")
  dir.create(tmp, recursive = TRUE, showWarnings = FALSE)
  on.exit(unlink(tmp, recursive = TRUE), add = TRUE)
  expect_identical(nrow(datalib_countries(root = tmp)), 0L)
})
